controllers/component_image_controller.go

```diff
@@ -0,0 +1,248 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/types"
+	"knative.dev/pkg/apis"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/go-logr/logr"
+	appstudiov1alpha1 "github.com/redhat-appstudio/application-service/api/v1alpha1"
+	tektonapi "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+)
+
+const (
+	ComponentAnnotationName = "build.appstudio.openshift.io/component"
+	ComponentTaskLabelName  = "build.appstudio.openshift.io/component"
+	PipelineRunLabelName    = "tekton.dev/pipelineRun"
+	PipelineTaskLabelName   = "tekton.dev/pipelineTask"
+	BuildImageTaskName      = "build-container"
+)
+
+// NewComponentImageReconciler reconciles a Frigate object
+type NewComponentImageReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	Log    logr.Logger
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *NewComponentImageReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&tektonapi.PipelineRun{}, builder.WithPredicates(predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool {
+				return false
+			},
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				// React only if pipeline build has just finished successfully and a new image is ready
+
+				var old *tektonapi.PipelineRun
+				var new *tektonapi.PipelineRun
+				old, ok := e.ObjectOld.(*tektonapi.PipelineRun)
+				if !ok {
+					return false
+				}
+				new, ok = e.ObjectNew.(*tektonapi.PipelineRun)
+				if !ok {
+					return false
+				}
+
+				if new.Status.CompletionTime == nil || new.Status.CompletionTime.IsZero() {
+					// Pipeline run is still in progress
+					return false
+				}
+
+				if old.Status.CompletionTime != nil {
+					// Build had been finished before this update
+					return false
+				}
+
+				// Ensure the PipelineRun belongs to a Component
+				if new.ObjectMeta.Annotations[ComponentAnnotationName] == "" && new.ObjectMeta.Labels[ComponentAnnotationName] == "" {
```

Comment:
for my own education @mmorhun  I looked at https://github.com/redhat-appstudio/application-service/blob/main/gitops/generate_build.go and I see that yeah the annotation gets set at https://github.com/redhat-appstudio/application-service/blob/86571e38f52e6bbe67b945dd58b695af26ae834d/gitops/generate_build.go#L87 as part of generating the trigger template

and the label gets set at https://github.com/redhat-appstudio/application-service/blob/86571e38f52e6bbe67b945dd58b695af26ae834d/gitops/generate_build.go#L87 for the "initial pipeline run"

given those separate paths, I'm "guessing" there is a chance where a problem could occur and one of the two might not be set, and we don't want to process in that case.

Is that the rationale? And that we can't assume if the label is there, the annotation will be, where if we could assume that, we would leverage a label selector during the construction of the cache when building the manager?